File: 2025/aoc/day10/factory_machine.py
class FactoryMachine:
    LIGHT_OFF = "."
    LIGHT_ON = "#"

    def __init__(self, uid, lights, buttons, joltages):
        self.__uid = uid
        # The desired state of the lights
        self.__lights_dia = lights

        # The actual state of the lights
        # Initially all off
        self.reset_lights()

        self.__buttons = buttons
        self.__joltages = joltages

    def __str__(self):
        return f"{self.__uid}) [{''.join(self.__lights_dia)}] {self.__buttons} {{{self.__joltages}}}"

    @property
    def button_count(self):
        return len(self.__buttons)

    @property
    def uid(self):
        return self.__uid

    @property
    def lights(self):
        return f"{self.__lights} / {self.__lights_dia}"

    def reset_lights(self):
        self.__lights = ["."] * len(self.__lights_dia)

    def lights_needed(self):
        needed = []

        for idx, state in enumerate(self.__lights_dia):
            if state == self.LIGHT_ON:
                needed.append(idx)

        return needed

    def lights_ready(self):
        return self.__lights == self.__lights_dia

    def lights_diff(self):
        ldiff = []

        for idx, light in enumerate(self.__lights):
            if light == self.__lights_dia[idx]:
                ldiff.append(True)
            else:
                ldiff.append(False)

        return ldiff

    def __toggle_light(self, lnum):
        self.__lights[lnum] = self.LIGHT_OFF if self.__lights[lnum] == self.LIGHT_ON else self.LIGHT_ON

    def find_buttons(self, *lights):
        buttons = []

        for idx, btn in enumerate(self.__buttons):
            found = False
            for lnum in lights:
                if lnum in btn:
                    found = True
                    break

            if found:
                buttons.append(idx)

        return buttons

    def push_buttons(self, *btns):
        for bnum in btns:
            for lnum in self.__buttons[bnum]:
                self.__toggle_light(lnum)

File: 2025/agoc/src/caroon.org/aoc25/day10/factory_machine.go
package day10

import (
	"bytes"
)

type FactoryMachine struct {
	lightsDia []byte
	lights    []byte
	buttons   [][]int
	joltages  []int
}

const LightOff byte = '.'
const LightOn byte = '#'

func NewFactoryMachine(lightsDia []byte, buttons [][]int, joltages []int) *FactoryMachine {
	var lightCount int = len(lightsDia)
	var machine FactoryMachine = FactoryMachine{
		lightsDia,
		make([]byte, lightCount),
		buttons,
		joltages,
	}

	machine.ResetLights()

	return &machine
}

func (fm *FactoryMachine) ButtonCount() int {
	return len(fm.buttons)
}

func (fm *FactoryMachine) ResetLights() {
	for idx := 0; idx < len(fm.lights); idx++ {
		fm.lights[idx] = LightOff
	}
}

func (fm *FactoryMachine) LightsReady() bool {
	return bytes.Compare(fm.lights, fm.lightsDia) == 0
}

func (fm *FactoryMachine) ToggleLight(lightNum int) {
	if fm.lights[lightNum] == LightOff {
		fm.lights[lightNum] = LightOn
	} else {
		fm.lights[lightNum] = LightOff
	}
}

func (fm *FactoryMachine) PushButtons(btns []int) {
	for _, bNum := range btns {
		for _, lNum := range fm.buttons[bNum] {
			fm.ToggleLight(lNum)
		}
	}
}
